tests/deployment/test_client_server.py

```diff
@@ -337,3 +352,61 @@ def check_input_compression(model, fhe_circuit_compressed, is_torch, **compilati
         "Compressed input ciphertext's is not smaller than the uncompressed input ciphertext. Got "
         f"{compressed_size} bytes (compressed) and {uncompressed_size} bytes (uncompressed)."
     )
+
+
+ERROR_MSG_BAD_MODE = "Mode must be either 'inference' or 'training'"
+ERROR_MSG_NO_FHE_CIRCUIT = "Training FHE circuit does not exist."
+
+
+@pytest.mark.parametrize("n_bits", [2])
+@pytest.mark.parametrize(
+    "mode, fit_encrypted, error_message",
+    [
+        ("invalid_mode", True, ERROR_MSG_BAD_MODE),
+        ("INVALID_MODE", True, ERROR_MSG_BAD_MODE),
+        ("train", True, ERROR_MSG_BAD_MODE),
+        ("", True, ERROR_MSG_BAD_MODE),
+        (None, True, None),
+        ("inference", False, None),
+        ("inference", True, None),
+        ("training", False, ERROR_MSG_NO_FHE_CIRCUIT),
+        ("training", True, None),
+        (DeploymentMode.INFERENCE, False, None),
+        (DeploymentMode.TRAINING, False, ERROR_MSG_NO_FHE_CIRCUIT),
+        (DeploymentMode.TRAINING, True, None),
+    ],
+)
+def test_save_mode_handling(n_bits, fit_encrypted, mode, error_message):
+    """Test that the save method handles valid and invalid modes correctly."""
+
+    # Generate random data
+    x, y = numpy.random.rand(20, 2), numpy.random.randint(0, 2, 20)
+
+    x_train = x[:-1]
+    y_train = y[:-1]
+
+    # Instantiate the model
+    parameters_range = [-1, 1] if fit_encrypted else None
+    model = instantiate_model_generic(
+        partial(SGDClassifier, fit_encrypted=fit_encrypted, parameters_range=parameters_range),
+        n_bits=n_bits,
+    )
+
+    # Fit the model
+    model.fit(x_train, y_train)
+
+    # Compile
+    model.compile(X=x_train)
+
+    # Create FHEModelDev instance
+    with tempfile.TemporaryDirectory() as temp_dir:
+        model_dev = FHEModelDev(path_dir=temp_dir, model=model)
+
+        if error_message:
+            with pytest.raises((AssertionError, ValueError), match=error_message):
+                model_dev.save(mode=mode)
+        else:
+            try:
```

Comment:
better to remove this try/except so that we properly catch the error is there is one no ? worst case, could you at least print the traceback using something like `except ValueError as e:` (and include e in the `pytest.fail`)

Reply:
Ah yes this was for a previous version of the test useless now. Thanks!

Reply:
done